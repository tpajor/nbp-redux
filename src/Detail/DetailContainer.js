import React from 'react';
import Clock from 'react-live-clock';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import { getInputedNumberOfLastRatesRequest, signInOrOut } from '../Actions';

import Input from './Input';
import Chart from './Chart';

import './DetailContainer.css';

class DetailContainer extends React.Component {
  state = { 
    contWidth: 0,
  };

  componentDidMount = () => {
    this.setWidth();
    window.addEventListener('resize', this.setWidth);
  };

  componentWillUnmount = () => {
    window.removeEventListener("resize", this.setWidth);
  };

  setWidth = () => {
    const width = document.getElementById('Chart').clientWidth;
    this.setState({ contWidth: width * 0.8, });
  };

  render() {
    const { contWidth } = this.state;
    return (
      <div  className={'col-8 DetailView'}>
        <Clock format={'YYYY-MM-DD HH:mm:ss'} ticking={true} timezone={'Europe/Warsaw'} />
        {this.props.detailViewData.signedIn ? 
          <button className="btn btn-primary SignButton" onClick={() => this.props.signInOrOut(false)}>
            Wyloguj
          </button> :
          <Link to='/login' className="btn btn-primary SignButton">
            Zaloguj
          </Link>
        }
        <div id={'Chart'} className={'Chart'}>
          <Input currencyCodeToViewInDetail={this.props.detailViewData.currencyCodeToViewInDetail}
            getInputedNumberOfLastRates={this.props.getInputedNumberOfLastRates}
            detailsInputError={this.props.detailViewData.detailsInputError}
            detailsInputErrorMessage={this.props.detailViewData.detailsInputErrorMessage}
          />
          <Chart dataToPlot={this.props.detailViewData.lastRates}
            contWidth={contWidth}
          />
        </div>
      </div>
    );
  }
};

const mapDispatchToProps = {
  getInputedNumberOfLastRates: getInputedNumberOfLastRatesRequest,
  signInOrOut,
};

DetailContainer.propTypes = {
  detailViewData: PropTypes.object,
};

export default connect(null, mapDispatchToProps)(DetailContainer);