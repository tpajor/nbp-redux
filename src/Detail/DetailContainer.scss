.DetailView {
  min-height: 100vh;
}

.Chart {
  height: auto;
  width: 70%;
  margin: 20px 20% 0 10%;
  padding: 0 10% 5% 5%;
  border: 1px solid #ccc;
  border-radius: 5px;
  box-shadow: 5px 5px #aaa;
}
