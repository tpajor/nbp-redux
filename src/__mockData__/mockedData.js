const card = { id: 'someId', code: 'USD', currency: 'dolar amerykański', rates: [{ ask: 3.3, bid: 3.4 }]};
const card2 = { id: 'someId2', code: 'USD', currency: 'dolar amerykański', rates: { ask: 3.3, bid: 3.4 }};

const search = {
  cards: [card, card2],
  temporaryCard: card,
  searchError: false,
  addCardError: true,
  addCardErrorMessage: 'error',
  currenciesTable: ['usd', 'chf', 'aud'],
  currencyCodeToViewInDetail: 'usd',
};

const user = { userName: 'User', pass: 'pass' };
const user2 = { userName: 'User2', pass: 'pass2'};

export const login = {
  signingError: true,
  signingErrorMessage: 'error',
  signedIn: true,
  userSignedIn: 'User',
  users: [user, user2]
};

const lastRates = [
  { ask: 3.5983, bid: 3.5271, effectiveDate: "2018-05-07", no: "087/C/NBP/2018" },
  { ask: 3.6088, bid: 3.5374, effectiveDate: "2018-05-08", no: "088/C/NBP/2018" }
];

const detail = {
  lastRates: lastRates,
  detailsInputError: false,
  detailsInputErrorMessage: '',
};

export const initialState = { search, login, detail };

export const searchViewData = {
  cards: search.cards,
  temporaryCard: search.temporaryCard,
  searchError: search.searchError,
  signedIn: login.signedIn,
  userSignedIn: login.userSignedIn,
  addCardError: search.addCardError,
  addCardErrorMessage: search.addCardErrorMessage,
  currenciesTable: search.currenciesTable,
};

export const detailViewData = {
  currencyCodeToViewInDetail: search.currencyCodeToViewInDetail,
  lastRates: detail.lastRates,
  detailsInputError: detail.detailsInputError,
  detailsInputErrorMessage: detail.detailsInputErrorMessage,
  signedIn: login.signedIn,
};