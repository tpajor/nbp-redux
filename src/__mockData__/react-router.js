const Link = () => {
  return jest.fn();
}

export default Link;