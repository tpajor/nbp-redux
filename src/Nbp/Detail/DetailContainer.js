import React from 'react';
import Clock from 'react-live-clock';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import { getInputedNumberOfLastRatesRequest } from './DetailActions';
import { signInOrOut } from '../../Login/LoginActions';

import Input from './Input';
import Chart from './Chart';

import './DetailContainer.css';

export class DetailContainer extends React.Component {
  state = { 
    contWidth: 0,
  };

  componentDidMount = () => {
    this.setWidth();
    window.addEventListener('resize', this.setWidth);
  };

  componentWillUnmount = () => {
    window.removeEventListener("resize", this.setWidth);
  };

  setWidth = () => {
    let width = 300;
    if(document.getElementById('Chart')) width = document.getElementById('Chart').clientWidth;
    this.setState({ contWidth: width * 0.8, });
  };

  render() {
    const { contWidth } = this.state;
    const { signInOrOut, getInputedNumberOfLastRates, detailViewData,  } = this.props;
    return (
      <div  className={'col-8 DetailView'}>
        <Clock format={'YYYY-MM-DD HH:mm:ss'} ticking={true} timezone={'Europe/Warsaw'} />
        {detailViewData.signedIn ? 
          <Link to='/login' className="btn btn-primary SignButton" onClick={() => signInOrOut(true)}>
            Wyloguj
          </Link> :
          <Link to='/login' className="btn btn-primary SignButton">
            Zaloguj
          </Link>
        }
        <div id={'Chart'} className={'Chart'}>
          <Input currencyCodeToViewInDetail={detailViewData.currencyCodeToViewInDetail}
            getInputedNumberOfLastRates={getInputedNumberOfLastRates}
            detailsInputError={detailViewData.detailsInputError}
            detailsInputErrorMessage={detailViewData.detailsInputErrorMessage}
          />
          <Chart dataToPlot={detailViewData.lastRates}
            contWidth={contWidth}
          />
        </div>
      </div>
    );
  }
};

const mapDispatchToProps = {
  getInputedNumberOfLastRates: getInputedNumberOfLastRatesRequest,
  signInOrOut,
};

DetailContainer.propTypes = {
  detailViewData: PropTypes.object,
  getInputedNumberOfLastRates: PropTypes.func,
  signInOrOut: PropTypes.func,
};

export default connect(null, mapDispatchToProps)(DetailContainer);