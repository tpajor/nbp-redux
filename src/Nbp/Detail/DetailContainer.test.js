import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { DetailContainer } from './DetailContainer';
import { detailViewData } from '../../__mockData__/mockedData';
import ReactRouterEnzymeContext from 'react-router-enzyme-context';

Enzyme.configure({ adapter: new Adapter() });

function setup() {
  const props = {
    signInOrOut: jest.fn(),
    getInputedNumberOfLastRates: jest.fn(),
    detailViewData: detailViewData
  };
  const enzymeWrapper = shallow(<DetailContainer {...props} />);

  return {
    enzymeWrapper,
    props
  };
};

describe('DetailContainer component', () => {
  it('should render self and subcomponents with their props', () => {
    const { enzymeWrapper } = setup();

    expect(enzymeWrapper.length).toBe(1);

    expect(enzymeWrapper.find('ReactLiveClock').length).toBe(1);

    expect(enzymeWrapper.find('div').length).toBe(2);
    expect(enzymeWrapper.find('div').first().hasClass('col-8 DetailView')).toBe(true);
    expect(enzymeWrapper.find('div').last().hasClass('Chart')).toBe(true);
    
    expect(enzymeWrapper.find('Link').length).toBe(1);
    expect(enzymeWrapper.find('Link').hasClass('btn btn-primary SignButton')).toBe(true);

    expect(enzymeWrapper.find('Input').length).toBe(1);
    expect(enzymeWrapper.find('Input').props().currencyCodeToViewInDetail).toEqual(detailViewData.currencyCodeToViewInDetail);
    expect(typeof enzymeWrapper.find('Input').props().getInputedNumberOfLastRates).toBe('function');
    expect(enzymeWrapper.find('Input').props().detailsInputError).toEqual(detailViewData.detailsInputError);
    expect(enzymeWrapper.find('Input').props().detailsInputErrorMessage).toEqual(detailViewData.detailsInputErrorMessage);

    expect(enzymeWrapper.find('Chart').length).toBe(1);
    expect(enzymeWrapper.find('Chart').props().dataToPlot).toEqual(detailViewData.lastRates);
    enzymeWrapper.setState({ contWidth: 100 });
    expect(enzymeWrapper.find('Chart').props().contWidth).toEqual(100);
  });

  it('should render correctly', () => {
    const { enzymeWrapper } = setup();
    
    expect(enzymeWrapper).toMatchSnapshot();
  });

  it('should set ContWidth in state', () => {
    const { props } = setup();
    const options = new ReactRouterEnzymeContext();
    const enzymeWrapper = mount(<DetailContainer {...props} />, options.get(), document.body, window );
    const instance = enzymeWrapper.instance();
    expect(enzymeWrapper.state(['contWidth'])).toEqual(240);
    const chart = document.createElement('Chart');
    chart.id = 'Chart';
    Object.defineProperty(chart, 'clientWidth', { value: 500 });
    document.body.appendChild(chart);
    instance.setWidth();
    expect(enzymeWrapper.state(['contWidth'])).toEqual(400);
  });

  it('should add event listener on window resize on component mount and delete it after dismounting', () => {
    global.window.addEventListener = jest.fn();
    const { enzymeWrapper } = setup();
    const instance = enzymeWrapper.instance();
    expect(global.window.addEventListener).toHaveBeenCalled();
    console.log(global.window.addEventListener);
    //expect(global.window.addEventListener).toHaveBeenLastCalledWith('resize', setWidth);
  });
});