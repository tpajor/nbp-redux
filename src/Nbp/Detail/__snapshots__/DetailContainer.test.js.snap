// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailContainer component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DetailContainer
    detailViewData={
        Object {
            "currencyCodeToViewInDetail": "usd",
            "detailsInputError": false,
            "detailsInputErrorMessage": "",
            "lastRates": Array [
              Object {
                "ask": 3.5983,
                "bid": 3.5271,
                "effectiveDate": "2018-05-07",
                "no": "087/C/NBP/2018",
              },
              Object {
                "ask": 3.6088,
                "bid": 3.5374,
                "effectiveDate": "2018-05-08",
                "no": "088/C/NBP/2018",
              },
            ],
            "signedIn": true,
          }
    }
    getInputedNumberOfLastRates={[Function]}
    signInOrOut={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ReactLiveClock
          className={null}
          date={null}
          format="YYYY-MM-DD HH:mm:ss"
          interval={1000}
          ticking={true}
          timezone="Europe/Warsaw"
/>,
        <Link
          className="btn btn-primary SignButton"
          onClick={[Function]}
          replace={false}
          to="/login"
>
          Wyloguj
</Link>,
        <div
          className="Chart"
          id="Chart"
>
          <Input
                    currencyCodeToViewInDetail="usd"
                    detailsInputError={false}
                    detailsInputErrorMessage=""
                    getInputedNumberOfLastRates={[Function]}
          />
          <Chart
                    contWidth={240}
                    dataToPlot={
                              Array [
                                        Object {
                                          "ask": 3.5983,
                                          "bid": 3.5271,
                                          "effectiveDate": "2018-05-07",
                                          "no": "087/C/NBP/2018",
                                        },
                                        Object {
                                          "ask": 3.6088,
                                          "bid": 3.5374,
                                          "effectiveDate": "2018-05-08",
                                          "no": "088/C/NBP/2018",
                                        },
                                      ]
                    }
          />
</div>,
      ],
      "className": "col-8 DetailView",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": null,
          "date": null,
          "format": "YYYY-MM-DD HH:mm:ss",
          "interval": 1000,
          "ticking": true,
          "timezone": "Europe/Warsaw",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Wyloguj",
          "className": "btn btn-primary SignButton",
          "onClick": [Function],
          "replace": false,
          "to": "/login",
        },
        "ref": null,
        "rendered": "Wyloguj",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Input
              currencyCodeToViewInDetail="usd"
              detailsInputError={false}
              detailsInputErrorMessage=""
              getInputedNumberOfLastRates={[Function]}
/>,
            <Chart
              contWidth={240}
              dataToPlot={
                            Array [
                                          Object {
                                            "ask": 3.5983,
                                            "bid": 3.5271,
                                            "effectiveDate": "2018-05-07",
                                            "no": "087/C/NBP/2018",
                                          },
                                          Object {
                                            "ask": 3.6088,
                                            "bid": 3.5374,
                                            "effectiveDate": "2018-05-08",
                                            "no": "088/C/NBP/2018",
                                          },
                                        ]
              }
/>,
          ],
          "className": "Chart",
          "id": "Chart",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "currencyCodeToViewInDetail": "usd",
              "detailsInputError": false,
              "detailsInputErrorMessage": "",
              "getInputedNumberOfLastRates": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "contWidth": 240,
              "dataToPlot": Array [
                Object {
                  "ask": 3.5983,
                  "bid": 3.5271,
                  "effectiveDate": "2018-05-07",
                  "no": "087/C/NBP/2018",
                },
                Object {
                  "ask": 3.6088,
                  "bid": 3.5374,
                  "effectiveDate": "2018-05-08",
                  "no": "088/C/NBP/2018",
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ReactLiveClock
            className={null}
            date={null}
            format="YYYY-MM-DD HH:mm:ss"
            interval={1000}
            ticking={true}
            timezone="Europe/Warsaw"
/>,
          <Link
            className="btn btn-primary SignButton"
            onClick={[Function]}
            replace={false}
            to="/login"
>
            Wyloguj
</Link>,
          <div
            className="Chart"
            id="Chart"
>
            <Input
                        currencyCodeToViewInDetail="usd"
                        detailsInputError={false}
                        detailsInputErrorMessage=""
                        getInputedNumberOfLastRates={[Function]}
            />
            <Chart
                        contWidth={240}
                        dataToPlot={
                                    Array [
                                                Object {
                                                  "ask": 3.5983,
                                                  "bid": 3.5271,
                                                  "effectiveDate": "2018-05-07",
                                                  "no": "087/C/NBP/2018",
                                                },
                                                Object {
                                                  "ask": 3.6088,
                                                  "bid": 3.5374,
                                                  "effectiveDate": "2018-05-08",
                                                  "no": "088/C/NBP/2018",
                                                },
                                              ]
                        }
            />
</div>,
        ],
        "className": "col-8 DetailView",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": null,
            "date": null,
            "format": "YYYY-MM-DD HH:mm:ss",
            "interval": 1000,
            "ticking": true,
            "timezone": "Europe/Warsaw",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Wyloguj",
            "className": "btn btn-primary SignButton",
            "onClick": [Function],
            "replace": false,
            "to": "/login",
          },
          "ref": null,
          "rendered": "Wyloguj",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Input
                currencyCodeToViewInDetail="usd"
                detailsInputError={false}
                detailsInputErrorMessage=""
                getInputedNumberOfLastRates={[Function]}
/>,
              <Chart
                contWidth={240}
                dataToPlot={
                                Array [
                                                Object {
                                                  "ask": 3.5983,
                                                  "bid": 3.5271,
                                                  "effectiveDate": "2018-05-07",
                                                  "no": "087/C/NBP/2018",
                                                },
                                                Object {
                                                  "ask": 3.6088,
                                                  "bid": 3.5374,
                                                  "effectiveDate": "2018-05-08",
                                                  "no": "088/C/NBP/2018",
                                                },
                                              ]
                }
/>,
            ],
            "className": "Chart",
            "id": "Chart",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "currencyCodeToViewInDetail": "usd",
                "detailsInputError": false,
                "detailsInputErrorMessage": "",
                "getInputedNumberOfLastRates": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "contWidth": 240,
                "dataToPlot": Array [
                  Object {
                    "ask": 3.5983,
                    "bid": 3.5271,
                    "effectiveDate": "2018-05-07",
                    "no": "087/C/NBP/2018",
                  },
                  Object {
                    "ask": 3.6088,
                    "bid": 3.5374,
                    "effectiveDate": "2018-05-08",
                    "no": "088/C/NBP/2018",
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
